File: src/com/timepath/hl2/gameinfo/Player.java
package com.timepath.hl2.gameinfo;

import java.util.LinkedList;
import java.util.List;
import java.util.logging.Logger;

/**
 *
 * @author TimePath
 */
public class Player {

    private static final Logger LOG = Logger.getLogger(Player.class.getName());

    /**
     * Makes players enemies
     *
     * Function:
     * Adds a new enemy.
     * Makes your enemy's enemies your allies.
     * Makes your enemy's allies your enemies.
     * Informs all your allies of your new enemy.
     *
     * @param v Victim
     * @param k Killer
     */
    static void exchangeInfo(Player v, Player k) {
        List<Player> vAllies = new LinkedList<>(v.getAllies());
        List<Player> vEnemies = new LinkedList<>(v.getEnemies());

        List<Player> kAllies = new LinkedList<>(k.getAllies());
        List<Player> kEnemies = new LinkedList<>(k.getEnemies());

        if(k.getAllies().contains(v) || v.getAllies().contains(k)) { // Traitor
            for(int i = 0; i < k.getAllies().size(); i++) {
                k.getAllies().get(i).getAllies().remove(k);

                k.getAllies().get(i).getEnemies().remove(k);
                k.getAllies().get(i).getEnemies().add(k);
            }
            for(int i = 0; i < v.getAllies().size(); i++) {
                v.getAllies().get(i).getAllies().remove(k);

                v.getAllies().get(i).getEnemies().remove(k);
                v.getAllies().get(i).getEnemies().add(k);
            }

            for(int i = 0; i < k.getEnemies().size(); i++) {
                k.getEnemies().get(i).getEnemies().remove(k);

                k.getEnemies().get(i).getAllies().remove(k);
                k.getEnemies().get(i).getAllies().add(k);
            }
            for(int i = 0; i < v.getEnemies().size(); i++) {
                v.getEnemies().get(i).getEnemies().remove(k);

                v.getEnemies().get(i).getAllies().remove(k);
                v.getEnemies().get(i).getAllies().add(k);
            }

            k.getAllies().clear();
            k.getEnemies().clear();

            k.getAllies().addAll(kEnemies);
//            k.allies.addAll(vEnemies);
            k.getEnemies().addAll(kAllies);
//            k.enemies.addAll(vAllies);
        } else {
            v.addEnemy(k);
            k.addEnemy(v);

            v.addAllies(kEnemies);
            v.addEnemies(kAllies);

            k.addAllies(vEnemies);
            k.addEnemies(vAllies);
        }
    }

    private final List<Player> allies = new LinkedList<>();

    private final List<Player> enemies = new LinkedList<>();

    private String name;

    Player(String name) {
        this.name = name;
    }

    public void addAllies(List<Player> list) {
        for(Player ally : list) {
            addAllyR(ally);
            for(Player enemy : getEnemies()) {
                enemy.addEnemyR(ally);
            }
        }
    }
    public void addAlly(Player other) {
        if(other == this) {
            return;
        }
        if(!allies.contains(other)) {
            getAllies().add(other);
        }
        if(!other.hasAlly(this)) {
            other.getAllies().add(this);
        }
    }

    public void addAllyR(Player other) {
        addAlly(other);
        for(int j = 0; j < getAllies().size(); j++) {
            getAllies().get(j).addAlly(other);
        }
    }

    public void addEnemies(List<Player> enemies) {
        for(Player list1 : enemies) {
            addEnemyR(list1);
            for(Player ally : getAllies()) {
                ally.addEnemyR(list1);
            }
        }
    }

    public void addEnemy(Player other) {
        if(other == this) {
            return;
        }
        if(!enemies.contains(other)) {
            getEnemies().add(other);
        }
        if(!other.hasEnemy(this)) {
            other.getEnemies().add(this);
        }
    }

    public void addEnemyR(Player other) {
        addEnemy(other);
        for(int j = 0; j < getAllies().size(); j++) {
            getAllies().get(j).addEnemy(other);
        }
    }

    /**
     * @return the allies
     */public List<Player> getAllies() {
         return allies;
     }

    /**
     * @return the enemies
     */
    public List<Player> getEnemies() {
        return enemies;
    }

    /**
     * @return the name
     */
    public String getName() {
        return name;
    }

    /**
     * @param name the name to set
     */
    public void setName(String name) {
        this.name = name;
    }

    //<editor-fold defaultstate="collapsed" desc="toString()">

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(getName());
        
        sb.append(" (e:{");
        List<Player> myEnemies = getEnemies();
        for(int i = 0; i < getEnemies().size(); i++) {
            sb.append(myEnemies.get(i).getName());
            if(i + 1 < getEnemies().size()) {
                sb.append(", ");
            }
        }
        sb.append("}, a:{");
        List<Player> myAllies = getAllies();
        for(int i = 0; i < getAllies().size(); i++) {
            sb.append(myAllies.get(i).getName());
            if(i + 1 < getAllies().size()) {
                sb.append(", ");
            }
        }
        sb.append("})");
        
        return sb.toString();
    }
    //</editor-fold>
    
    private boolean hasAlly(Player player) {
        return allies.contains(player);
    }

    private boolean hasEnemy(Player player) {
        return enemies.contains(player);
    }

}

File: src/main/kotlin/com/timepath/hl2/gameinfo/Player.kt
package com.timepath.hl2.gameinfo

import java.util.LinkedList
import java.util.logging.Logger

/**
 * @author TimePath
 */
public class Player(public var name: String?) {
    val allies = LinkedList<Player>()
    val enemies = LinkedList<Player>()

    fun addAllies(list: Iterable<Player>) {
        for (ally in list) {
            addAllyR(ally)
            for (enemy in enemies) {
                enemy.addEnemyR(ally)
            }
        }
    }

    fun addAlly(other: Player) {
        if (other == this) {
            return
        }
        if (other !in allies) {
            allies.add(other)
        }
        if (!other.hasAlly(this)) {
            other.allies.add(this)
        }
    }

    fun addAllyR(other: Player) {
        addAlly(other)
        for (ally in allies) {
            ally.addAlly(other)
        }
    }

    fun addEnemies(enemies: Iterable<Player>) {
        for (list1 in enemies) {
            addEnemyR(list1)
            for (ally in allies) {
                ally.addEnemyR(list1)
            }
        }
    }

    fun addEnemy(other: Player) {
        if (other == this) {
            return
        }
        if (other !in enemies) {
            enemies.add(other)
        }
        if (!other.hasEnemy(this)) {
            other.enemies.add(this)
        }
    }

    fun addEnemyR(other: Player) {
        addEnemy(other)
        for (ally in allies) {
            ally.addEnemy(other)
        }
    }

    override fun toString(): String {
        val sb = StringBuilder()
        sb.append(name)
        sb.append(" (e:{")
        val myEnemies = enemies
        for (i in enemies.indices) {
            sb.append(myEnemies[i].name)
            if ((i + 1) < enemies.size()) {
                sb.append(", ")
            }
        }
        sb.append("}, a:{")
        val myAllies = allies
        for (i in allies.indices) {
            sb.append(myAllies[i].name)
            if ((i + 1) < allies.size()) {
                sb.append(", ")
            }
        }
        sb.append("})")
        return sb.toString()
    }

    private fun hasAlly(player: Player) = player in allies

    private fun hasEnemy(player: Player) = player in enemies

    companion object {

        private val LOG = Logger.getLogger(javaClass<Player>().getName())

        /**
         * Makes players enemies
         * Function:
         * Adds a new enemy.
         * Makes your enemy's enemies your allies.
         * Makes your enemy's allies your enemies.
         * Informs all your allies of your new enemy.
         *
         * @param v
         *         Victim
         * @param k
         *         Killer
         */
        public fun exchangeInfo(v: Player, k: Player) {
            val vAllies = LinkedList(v.allies)
            val vEnemies = LinkedList(v.enemies)
            val kAllies = LinkedList(k.allies)
            val kEnemies = LinkedList(k.enemies)
            if (v in k.allies || k in v.allies) {
                // Traitor
                for (kAlly in k.allies) {
                    kAlly.allies.remove(k)
                    kAlly.enemies.remove(k)
                    kAlly.enemies.add(k)
                }
                for (vAlly in v.allies) {
                    vAlly.allies.remove(k)
                    vAlly.enemies.remove(k)
                    vAlly.enemies.add(k)
                }
                for (kEnemy in k.enemies) {
                    kEnemy.enemies.remove(k)
                    kEnemy.allies.remove(k)
                    kEnemy.allies.add(k)
                }
                for (vEnemy in v.enemies) {
                    vEnemy.enemies.remove(k)
                    vEnemy.allies.remove(k)
                    vEnemy.allies.add(k)
                }
                k.allies.clear()
                k.enemies.clear()
                k.allies.addAll(kEnemies)
                //            k.allies.addAll(vEnemies);
                k.enemies.addAll(kAllies)
                //            k.enemies.addAll(vAllies);
            } else {
                v.addEnemy(k)
                k.addEnemy(v)
                v.addAllies(kEnemies)
                v.addEnemies(kAllies)
                k.addAllies(vEnemies)
                k.addEnemies(vAllies)
            }
        }
    }
}
